iree/compiler/JitEval/Runtime.cpp

```diff
@@ -0,0 +1,232 @@
+      iree_hal_buffer_t* buffer = iree_hal_buffer_view_buffer(bufferView);
+
+      // Map the memory and construct.
+      DenseElementsAttr elementsAttr;
+      iree_hal_buffer_mapping_t mapping;
+      IREE_CHECK_OK(
+          iree_hal_buffer_map_range(buffer, IREE_HAL_MEMORY_ACCESS_READ,
+                                    /*byte_offset=*/0, length, &mapping));
+      ArrayRef<char> rawBufferArray(
+          reinterpret_cast<char*>(mapping.contents.data),
+          mapping.contents.data_length);
+      bool detectedSplat = false;
+      if (DenseElementsAttr::isValidRawBuffer(tensorType, rawBufferArray,
+                                              detectedSplat)) {
+        elementsAttr = DenseElementsAttr::getFromRawBuffer(
+            tensorType, rawBufferArray, detectedSplat);
+      }
+      iree_hal_buffer_unmap_range(&mapping);
+
+      if (!elementsAttr) {
+        emitError(loc) << "mapped memory region was not valid for constructing "
+                          "tensor of type "
+                       << tensorType << " (length=" << length << ")";
+      }
+      return elementsAttr;
+    } else {
+      iree_string_view_t typeName =
+          iree_vm_ref_type_name(variant.type.ref_type);
+      emitError(loc) << "unrecognized evaluated ref type: "
+                     << StringRef(typeName.data, typeName.size);
+      return {};
+    }
+  }
+
+  emitError(loc) << "unrecognized evaluated variant type";
+  return {};
+}
+
+void CompiledBinary::initialize(void* data, size_t length) {
+  Runtime& runtime = Runtime::getInstance();
+  IREE_CHECK_OK(iree_vm_bytecode_module_create(
+      iree_make_const_byte_span(data, length), iree_allocator_null(),
+      iree_allocator_system(), &main_module));
+
+  std::array<iree_vm_module_t*, 2> modules = {runtime.hal_module, main_module};
+  IREE_CHECK_OK(iree_vm_context_create_with_modules(
+      instance, IREE_VM_CONTEXT_FLAG_NONE, modules.data(), modules.size(),
+      iree_allocator_system(), &context));
+}
+
+LogicalResult InMemoryCompiledBinary::translateFromModule(
+    mlir::ModuleOp moduleOp) {
+  llvm::raw_string_ostream os(binary);
+  iree_compiler::IREE::VM::BytecodeTargetOptions targetOptions;
+  if (failed(iree_compiler::IREE::VM::translateModuleToBytecode(
+          moduleOp, targetOptions, os))) {
+    return failure();
+  }
+  os.flush();
+  initialize(&binary[0], binary.length());
+  return success();
+}
+
+Runtime::Runtime() {
+  registry = iree_hal_driver_registry_default();
+  IREE_CHECK_OK(iree_hal_vmvx_driver_module_register(registry));
+  IREE_CHECK_OK(iree_hal_module_register_types());
+
+  // Create driver and device.
+  iree_hal_driver_t* driver = nullptr;
+  IREE_CHECK_OK(iree_hal_driver_registry_try_create_by_name(
+      registry, iree_make_cstring_view("vmvx"), iree_allocator_system(),
+      &driver));
+  IREE_CHECK_OK(iree_hal_driver_create_default_device(
+      driver, iree_allocator_system(), &device));
```

Comment:
The device and the hal module are not thread-safe and they also carry some state that could bleed across runs - they're extremely cheap to create though so doing that per use is fine. The driver is the expensive thing (owns the thread pool) and is thread-safe.

Reply:
sg